File: 341.cpp
#include<iostream>
#include<vector>
#include<stack>

using namespace std;
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 */
 class NestedInteger {
   public:
     // Return true if this NestedInteger holds a single integer, rather than a nested list.
     bool isInteger() const;

     // Return the single integer that this NestedInteger holds, if it holds a single integer
     // The result is undefined if this NestedInteger holds a nested list
     int getInteger() const;

     // Return the nested list that this NestedInteger holds, if it holds a nested list
     // The result is undefined if this NestedInteger holds a single integer
     const vector<NestedInteger> &getList() const;
 };
class NestedIterator {
public:
    stack<NestedInteger> s;
    NestedIterator(vector<NestedInteger> &nestedList) {
        for(int i=nestedList.size()-1;i>=0;--i){
            s.push(nestedList[i]);
        }
    }
    
    int next() {
        int result=s.top().getInteger();
        s.pop();
        return result;
    }
    
    bool hasNext() {
        while(!s.empty()){
            NestedInteger tmp=s.top();
            if(tmp.isInteger())
                return true;
            s.pop();
            vector<NestedInteger>& next=tmp.getList();
            for(int i=next.size()-1;i>=0;--i){
                s.push(next[i]);
            }
        }
        return false;
    }
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */

File: Python/341.py
#dfs
class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.elems = []
        def dfs(ls):
            for l in ls:
                if l.isInteger():
                    self.elems.append(l.getInteger())
                else:
                    dfs(l.getList())
        dfs(nestedList)
        self.pos = -1

    def next(self):
        """
        :rtype: int
        """
        self.pos +=1
        return self.elems[self.pos]


    def hasNext(self):
        """
        :rtype: bool
        """
        return self.pos < len(self.elems) -1


#stack
class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.elems = []
        for i in range(len(nestedList)-1, -1, -1):
            self.elems.append(nestedList[i])


    def next(self):
        """
        :rtype: int
        """
        l = self.elems.pop()
        return l.getInteger()


    def hasNext(self):
        """
        :rtype: bool
        """
        while self.elems:
            l = self.elems[-1]
            if l.isInteger():
                return True
            self.elems.pop()
            ls = l.getList()
            for i in range(len(ls)-1, -1, -1):
                self.elems.append(ls[i])
        return False
